src/llvm-propagate-addrspaces.cpp

```diff
@@ -74,67 +77,136 @@ static bool isSpecialAS(unsigned AS) {
     return AddressSpace::FirstSpecial <= AS && AS <= AddressSpace::LastSpecial;
 }
 
+void PropagateJuliaAddrspaces::PoisonValues(std::vector<Value *> &Worklist) {
+    while (!Worklist.empty()) {
+        Value *CurrentV = Worklist.back();
+        Worklist.pop_back();
+        for (Value *User : CurrentV->users()) {
+            if (Visited.count(User))
+                continue;
+            Visited.insert(CurrentV);
+            Worklist.push_back(User);
+        }
+    }
+}
+
 Value *PropagateJuliaAddrspaces::LiftPointer(Value *V, Type *LocTy, Instruction *InsertPt) {
     SmallVector<Value *, 4> Stack;
-    Value *CurrentV = V;
+    std::vector<Value *> Worklist;
+    std::set<Value *> LocalVisited;
+    Worklist.push_back(V);
     // Follow pointer casts back, see if we're based on a pointer in
     // an untracked address space, in which case we're allowed to drop
     // intermediate addrspace casts.
-    while (true) {
-        Stack.push_back(CurrentV);
-        if (isa<BitCastInst>(CurrentV))
-            CurrentV = cast<BitCastInst>(CurrentV)->getOperand(0);
-        else if (isa<AddrSpaceCastInst>(CurrentV)) {
-            CurrentV = cast<AddrSpaceCastInst>(CurrentV)->getOperand(0);
-            if (!isSpecialAS(getValueAddrSpace(CurrentV)))
-                break;
+    while (!Worklist.empty()) {
+        Value *CurrentV = Worklist.back();
+        Worklist.pop_back();
+        if (LocalVisited.count(CurrentV)) {
+            continue;
         }
-        else if (isa<GetElementPtrInst>(CurrentV)) {
-            if (LiftingMap.count(CurrentV)) {
-                CurrentV = LiftingMap[CurrentV];
+        while (true) {
+            if (isa<BitCastInst>(CurrentV))
```

Comment:
`dyn_cast` would be more conventional (do LLVM developers care much about these details? we should try to get this pass merged upstream)

Reply:
Fair enough, though the way to get this upstream would be to work with the LLVM people on figuring out how to tell the regular optimizer that these transformations are legal.